src/QirRuntime/README.md

```diff
@@ -55,35 +49,53 @@ Running cmake from the editors will likely default to MSVC or clang-cl and fail.
 
 See [https://code.visualstudio.com/docs/remote/wsl] on how to use VS Code with WSL.
 
-## Test
+#### Other Prerequisites
+The build depends on `Microsoft.Quantum.Simulator.Runtime` dynamic library built at a higher level of the directory tree.
+To build that library follow the instructions in [`qsharp-runtime/README.md`](../../README.md#building-from-source)
+(up to and including the step `Simulation.sln`).
 
-Some of the tests depend on Microsoft.Quantum.Simulator.Runtime library. To run them make sure to build Native simulator
- from this repository or provide your own version of the library in a folder the OS would search during dynamic library
- lookup.
 
-Some of the tests use generated QIR (*.ll) files as build input. Currently the files are checked-in as part of the project
- but in the future they will be replaced by automatic generation during build. To regenerate the files, run generateqir.py
- or build/test scripts without specifying `noqirgen`. To use the checked-in files without regenerating them, run build/test
- scripts with `noqirgen` argument.
+### Build Commands
+To build QirRuntime you can run [`build-qir-runtime.ps1`](build-qir-runtime.ps1) script from QirRuntime folder:
+```batch
+pwsh build-qir-runtime.ps1
+```
+  
+The script will create the `build/{Debug|Release}` folder and place the build artifacts in it. The configuration `Debug|Release`
+is specified with the `BUILD_CONFIGURATION` environment variable.
+If the variable is not set then the default is specified in [`set-env.ps1`](../../build/set-env.ps1).  
+We strongly recommend doing local builds using this [`build-qir-runtime.ps1`](build-qir-runtime.ps1) build script 
+because it is used by the continuous integration (CI) infrastructure, and this script also runs clang-tidy if it is installed.
 
-### Running tests with test.py
+Or you can use CMake directly (not recommended). For example in Windows you can enter the following commands in the command line:
+```batch
+@rem Navigate to QirRuntime folder.
 
-To execute all tests locally run `test.py` from the project's root folder:
+@rem Creatge the `build` directory:
+mkdir build
 
-- (Windows) `python test.py [nobuild] [debug/release]`
-- (Linux) `python3 test.py [nobuild] [debug/release]`
+@rem Enter the `build` directory:
+pushd build
 
-The script will trigger an incremental build unless `nobuild` options is specified. Tests from the "[skip]" category
- won't be run.
+@rem Prepare for the build:
+cmake -G Ninja -DCMAKE_BUILD_TYPE=Debug ..
 
-### Running tests with CTest
+@rem Build (you will likely need to issue this command several times, see issue #589):
+cmake --build .
```

Comment:
I still don't understand this. Why does this need to be run multiple times? I haven't observed that behavior, though I'm trying now on a fresh repo to see if there is something I'm missing.

Reply:
I saw similar behavior during multithreaded builds ( `make -j ..` ). Different targets are built by different threads. And sometimes the later targets (that depend on the results of the earlier target builds) start earlier than the earlier target build  fully completes. It felt like the earlier target step was completed (from the point of view of the build utility), but the file output of the earlier stage was cached for a few seconds (e.g. the object file was not written to the disk completely), but the linking stage was already kicked off, and the linker was failing to open the object file, and the linking was failing.
At that moment the cached file was flushed onto the disk and the repeated build attempt was succeeding.

Reply:
I haven't observed this behavior with CMake, which does not perform parallel builds by default. In fact, CMake will build in dependency order based on detected dependencies across different targets, with the order of the folders in the top level CMakeLists.txt being the fallback. This ensures that CMake will only need to be run once. Have you gotten this error in your local builds?

I did just try it again, and after running bootstrap.ps1 and copying the relevant .ll files to the correct locations, I was able to manually create `src\QirRuntime\build\Debug` and then run these two commands to successfully build the whole QIR runtime and tests:
```
cmake -G Ninja ../..
cmake --build . --target install
```

Reply:
Also worth noting, the `--target install` ensures that all the binary output is collected in the `bin` folder, which the test script expects.

Reply:
I just removed those steps.

Reply:
I double checked out of curiosity. Looks like `--target install` was the missing part.